src/vsr/replica.zig

```diff
@@ -1400,46 +1445,92 @@ pub fn ReplicaType(
+                // We were already in this view prior to receiving the SV.
+                assert(self.status == .normal or self.status == .recovering_head);
 
-            assert(self.status == .view_change);
-            assert(message.header.view == self.view);
-            assert(message.header.op == op_highest(message_body_as_headers(message)));
+                log.debug("{}: on_start_view view={} (ignoring, old message)", .{
+                    self.replica,
+                    self.log_view,
+                });
+                return;
+            }
 
-            self.set_op_and_commit_max(message.header.op, message.header.commit, "on_start_view");
-            for (message_body_as_headers_chain_consecutive(message)) |*header| {
-                self.replace_header(header);
+            if (self.status == .recovering_head) {
+                self.view = message.header.view;
+            } else {
+                self.view_jump(message.header);
+                assert(self.view == message.header.view);
+
+                if (self.status == .normal) {
+                    assert(self.backup());
+                    assert(self.view == self.log_view);
+                }
             }
 
-            assert(self.op == message.header.op);
+            const view_headers = message_body_as_headers_chain_disjoint(message);
+            vsr.Headers.ViewChangeSlice.verify(view_headers);
 
-            assert(self.status == .view_change);
-            self.transition_to_normal_from_view_change_status(message.header.view);
-            self.send_prepare_oks_after_view_change();
+            for (view_headers) |*header| {
+                if (header.op <= self.op_checkpoint_trigger()) {
+                    if (self.log_view < self.view or
+                        (self.log_view == self.view and header.op >= self.op))
+                    {
+                        self.set_op_and_commit_max(header.op, message.header.commit, "on_start_view");
+                        assert(self.op == header.op);
+                        assert(self.commit_max >= message.header.commit);
+                        break;
+                    }
+                }
+            } else {
+                // This replica is too far behind, i.e. the new `self.op` is too far ahead of
+                // the last checkpoint. If we wrap now, we overwrite un-checkpointed transfers
+                // in the WAL, precluding recovery.
+                // TODO State transfer.
+                @panic("unimplemented (state transfer)");
+            }
+
+            self.view_headers.replace(view_headers);
+            assert(self.view_headers.array.get(0).view <= self.view);
+            assert(self.view_headers.array.get(0).op == message.header.op);
+            maybe(self.view_headers.array.get(0).op > self.op_checkpoint_trigger());
+            vsr.Headers.ViewChangeSlice.verify(self.view_headers.array.constSlice());
+
+            for (view_headers) |*header| {
```

Comment:
```suggestion
            for (self.view_headers.array) |*header| {
```

As we installed the headers, let's use the installed version.